README.md
#Vincent

[![Travs-CI status](https://travis-ci.org/wrobstory/vincent.png)](https://travis-ci.org/wrobstory/vincent)

![Vincent](http://farm9.staticflickr.com/8521/8644902478_0d1513db92_o.jpg)

###A Python to Vega translator

The folks at Trifacta are making it easy to build visualizations on top of D3 with Vega. Vincent makes it easy to build Vega with Python.  

Concept
-------
The data capabilities of Python. The visualization capabilities of JavaScript.

Vincent takes Python data structures (tuples, lists, dicts, and Pandas DataFrames) and translates them into [Vega](https://github.com/trifacta/vega) visualization grammar. It allows for quick iteration of visualization designs via simple addition and subtraction of grammar elements, and outputs the final visualization to JSON.

Getting Started
---------------

Lets build the Vega [bar chart example](https://github.com/trifacta/vega/wiki/Tutorial) with Vincent manually, then show some shortcuts to help get there a little faster, as well as methods for manipulating the components. First, create a Vincent object, which will initialize with some default parameters for the visualization: 
```python
>>> import vincent
>>> vis = vincent.Vega()
```
Now add some data. We could pass a dict, but nested tuples will keep our data sorted properly: 
```python
>>> vis.tabular_data((('A', 28), ('B', 55), ('C', 43), ('D', 91), ('E', 81), ('F', 53),
                     ('G', 19), ('H', 87), ('I', 52)))
```
Pass components to the visualization grammar as keyword arguments (skipping the 'marks' component here for brevity): 
```python
>>> vis.build_component(axes=[{"type":"x", "scale":"x"},{"type":"y", "scale":"y"}],
                        scales=[{"name":"x", "type":"ordinal", "range":"width", 
                                 "domain":{"data":"table", "field":"data.x"}},
                                {"name":"y", "range":"height", "nice":True, 
                                 "domain":{"data":"table", "field":"data.y"}}])
```
One option is to output to JSON:
```python
>>> vis.to_json(path)
```
Then copy/paste the JSON output to [Vega's online editor](http://trifacta.github.io/vega/editor/), where you should see a replica of the example. 

The other option is to output the Vega grammar and data into separate JSONs, as well as a simple [HTML scaffold](https://github.com/trifacta/vega/wiki/Runtime), then fire up a simple Python server locally: 

```python
>>> vis.to_json(path, split_data=True, html=True)
```
```
$ python -m SimpleHTTPServer 8000
```

CD to your path and point your browser at http://localhost:8000/vega_template.html to see your visualization. Please see the README in the examples folder for a more complete description of running the example files. 

Creating visualizations manually is a little tedious. The vincent.Vega() object can be subclassed to pre-define components and parameters. Lets take a shortcut and use the Bar class:  
```python
>>> import random
>>> vis = vincent.Bar()
>>> vis.tabular_data([random.randint(10, 100) for x in range(0, 21, 1)])
>>> vis.to_json(path)
```
![Bar](http://farm9.staticflickr.com/8532/8645065132_3f96e1be49.jpg)

Vincent also allows you to add and subtract components at varying levels of nesting depth in order to change the visualization. Vincent syntax for modifying component pieces on the fly is:
> Addition: += ( **New Value**, **Component**, **Component index**, **Keywords into nested structure** )

> Removal: -= ( **Old key**, **Component**, **Component index**, **Keywords into nested structure** ) 

For example, if we wanted to change the bar plot to an area plot: 
```python
>>> vis -= ('width', 'marks', 0, 'properties', 'enter') 
>>> vis += ('area', 'marks', 0, 'type')
>>> vis += ({'value': 'basis'}, 'marks', 0, 'properties', 'enter', 'interpolate')
>>> vis += ('linear', 'scales', 0, 'type')
>>> vis.to_json(path)
```

![Area](http://farm9.staticflickr.com/8540/8645065128_d2cf65bdf9_o.jpg)

You can create new visualizations from existing `Vega` objects with similar syntax
```python
>>> new_vis = vis + ('area', 'marks', 0, 'type')
>>> newer_vis = new_vis - ('scale', 'axes', 0)
```

Vincent also plays nice with Pandas DatetimeIndex, and allows for simple axis labeling and titling: 

```python
import vincent
import pandas as pd

# All of the following import code comes from Wes McKinney's book, Python for Data Analysis

import pandas.io.data as web
all_data = {}
for ticker in ['AAPL', 'GOOG']:
    all_data[ticker] = web.get_data_yahoo(ticker, '1/1/2010', '1/1/2013')
price = pd.DataFrame({tic: data['Adj Close']
                      for tic, data in all_data.iteritems()})

# Create line graph, with monthly plotting on the axes                       
line = vincent.Line()
line.tabular_data(price, columns=['AAPL'], axis_time='month')
line.to_json(path)

# Manipulate the axis tick/tick label orientation
line += ({'labels': {'angle': {'value': 25}}}, 'axes', 0, 'properties')
line += ({'value': 22}, 'axes', 0, 'properties', 'labels', 'dx')
line.update_vis(padding={'bottom': 50, 'left': 30, 'right': 30, 'top': 10})
line.update_vis(width=800, height=300)

# Add axis labels and a title
line.axis_label(y_label='AAPL Price', title='AAPL Stock Price 1/1/2010-1/1/2013')
line.to_json(path)
```
![AAPL fig](http://farm9.staticflickr.com/8393/8669181178_e22e576144_c.jpg)

Vincent fully supports the mapping component of Vega, allowing for rapid creation of regular maps and chloropleths by binding Python data structures to 
geoJSON data: 
```python
# Regular map, no data binding
vis = vincent.Map(width=1000, height=800)
vis.geo_data(projection='albersUsa', scale=1000, counties=r'county_geo.json')
vis += ('2B4ECF', 'marks', 0, 'properties', 'enter', 'stroke', 'value')
vis.geo_data(states=r'state_geo.json')
vis -= ('fill', 'marks', 1, 'properties', 'enter')
vis.to_json(path)
```
![Map](http://farm9.staticflickr.com/8389/8690908267_d7a3a83dae_z.jpg)
```python
# Bind Pandas DataFrame to map
vis = vincent.Map(width=1000, height=800)
vis.tabular_data(dataframe, columns=['FIPS_Code', 'Unemployment_rate_2011']) 
vis.geo_data(projection='albersUsa', scale=1000, bind_data='data.id',
             counties=r'county_geo.json')
vis += (["#f5f5f5","#000045"], 'scales', 0, 'range')
vis.to_json(path)
```
![Map](http://farm9.staticflickr.com/8543/8692026644_a1ee888398_z.jpg)

I also have Vincent incorporated into a [fork](https://github.com/wrobstory/d3py) of Mike Dewar's [d3py](https://github.com/mikedewar/d3py), with a pull request to merge into the main repo. 

For now, here is the syntax for using the d3py fork: 
```python
import d3py
import pandas as pd
import random

x = range(0, 21, 1)
y = [random.randint(25, 100) for num in range(0, 21, 1)]

df = pd.DataFrame({'x': x, 'y': y})

# Create Pandas figure
fig = d3py.PandasFigure(df, 'd3py_area', port=8080, columns=['x', 'y'])

# Add Vega Area plot
fig += d3py.vega.Area()

# Add interpolation to figure data
fig.vega += ({'value': 'basis'}, 'marks', 0, 'properties', 'enter', 
            'interpolate')
fig.show()
```

Dependencies
------------
pandas

requests (for shapefiles only)

Status
---------------
Note: Vincent is still in beta, but we are moving quickly towards a V1.0 release. Expect API/syntax to stay mostly fixed,
with major changes incorporated before V1.0. 


changelog.md
May 11 2012 V0.1.6, Rob Story, wrobstory

* iPython functionality no longer required import. Use ```import vincent.ipynb```



vincent.py
'''
Vincent
----

A Python to Vega translator.  

'''

from __future__ import print_function 
import json
import pandas as pd
import pdb

class Vega(object): 
    '''Vega abstract base class'''
    
    def __init__(self, width=400, height=200,
                 padding={'top': 10, 'left': 30, 'bottom': 20, 'right': 10},
                 viewport=None):
        '''
        The Vega classes generate JSON output in Vega grammer, a
        declarative format for creating and saving visualization designs.
        This class is meant to be an abstract base class on which to build
        the other piece of the complete VEGA specification. 
        
        A Vega object is initialized with only the Vega Visualization basic, 
        properties, with default values for the name, width, height, padding, 
        and viewport.
        
        Parameters:
        -----------

        width: int, default 800
            Width of the visualization
        height: int, default 400
            Height of the visualization
        padding: dict, default {'top': 10, 'left': 30, 'bottom': 20, 'right': 10}
            Internal margins for the visualization, Top, Left, Bottom, Right
        viewport: list, default None
            Width and height of on-screen viewport
            
        '''

        self.width = width
        self.height = height
        self.padding = padding
        self.viewport = viewport
        self.visualization = {'width': self.width, 
                              'padding': self.padding, 
                              'viewport': self.viewport}
        self.data = []
        self.scales = []
        self.axes = []
        self.marks = []
        self.build_vega()
        
    def __add__(self, tuple):
        '''Allow for updating of Vega with add operator'''
        self.update_component('add', *tuple)

    def __iadd__(self, tuple):
        '''Allow for updating of Vega with iadd operator'''
        self.update_component('add', *tuple)
        return self
        
    def __sub__(self, tuple): 
        '''Allow for updating of Vega with sub operator'''
        self.update_component('remove', *tuple)
        
    def __isub__(self, tuple): 
        '''Allow for updating of Vega with sub operator'''
        self.update_component('remove', *tuple)
        return self
        
    def build_vega(self, *args):
        '''Build complete vega specification. String arguments passed will not
        be included in vega dict. 
        
        Ex: object._build_vega('viewport')
        
        '''
        
        keys = ['width', 'height', 'padding', 'viewport', 'data', 
                'scales', 'axes', 'marks']
        self.vega = {}
        for key in keys: 
            if key not in args: 
                self.vega[key] = getattr(self, key)
                
    def update_vis(self, **kwargs):
        '''
        Update Vega Visualization basic properties:
        width, height, padding, viewport
        
        Ex: >>>my_vega.update_vis(height=800, width=800)
        '''
        
        for key, value in kwargs.iteritems():
            setattr(self, key, value)
            
        self.build_vega()
             
    def build_component(self, append=True, **kwargs):
        '''Build complete Vega component. The Vega grammar will update with
        passed keywords. This method defaults to appending to the existing 
        component; if you want to rebuild the component entirely, you must 
        pass append=False and the component as a list. 
        
        Examples:
        >>>my_vega.build_component(scales={"domain": {"data": "table",
                                                      "field": "data.x"},
                                           "name":"x", "type":"ordinal", 
                                           "range":"width"})
        >>>my_vega.build_component(axes=[{"scale": "x", type: "x"},
                                         {"scale": "y", type: "y"}], 
                                   append=False)
        
        '''

        for key, value in kwargs.iteritems(): 
            if append: 
                getattr(self, key).append(value)
            else: 
                setattr(self, key, value)
        
        self.build_vega()
        
    def update_component(self, change, value, parameter, index, *args):
        '''Update individual parameters of any component.
        
        Parameters: 
        -----------
        change: string, either 'add' or 'remove'
            'add' will add the value to the last specified key in *args (this
            can be a new key). 'remove' will remove the key specified by 
            'value'.
        value: Any JSON compatible datatype
            The value you want to substitute into the component
        parameter: string
            The Vega component you want to modify (scales, marks, etc)
        index: int
            The index of dict/object in the component array you want to mod
        
        Examples:
        >>>my_vega.update_component(add, 'w', 'axes', 0, 'scale')
        >>>my_vega.update_component('remove', 'width', 'marks', 0,
                                    'properties', 'enter')

        ''' 
        def set_keys(value, param, key, *args):
            if args: 
                return set_keys(value, param.get(key), *args)
            if change == 'add': 
                param[key] = value
            else:           
                param[key].pop(value)
    
        parameter = getattr(self, parameter)[index]
        if not args: 
            args = [value]
            if change == 'remove': 
                parameter.pop(value)
                self.build_vega()
                return 
        set_keys(value, parameter, *args)
        
        self.build_vega()
    
    def multi_update(self, comp_list): 
        '''Pass a list of component updates to change all'''
        
        for update in comp_list: 
            self.update_component(*update)
        
    def to_json(self, path):
        '''
        Save Vega object to JSON
        
        Parameters: 
        -----------
        path: string
            Save path
            
        '''
        
        with open(path, 'w') as f: 
            json.dump(self.vega, f, sort_keys=True, indent=4,
                      separators=(',', ': '))
                      
    def tabular_data(self, data, name="table", columns=None, use_index=False,
                     append=False):
        '''Create the data for a bar chart in Vega grammer. Data can be passed
        in a list, dict, or Pandas Dataframe. Note: old data is overwritten 
        
        Parameters:
        -----------
        name: string, default "table"
            Type of visualization
        columns: list, default None
            If passing Pandas DataFrame, you must pass at least one column name. 
            If one column is passed, x-values will default to the index values.
            If two column names are passed, x-values are columns[0], y-values 
            columns[1].
        use_index: boolean, default False
            Use the DataFrame index for your x-values
        append: boolean, default False
            Append new data to data already in vincent object
        
        Examples: 
        ---------
        >>>myvega.tabular_data([10, 20, 30, 40, 50])
        >>>myvega.tabular_data({'A': 10, 'B': 20, 'C': 30, 'D': 40, 'E': 50}
        >>>myvega.tabular_data(my_dataframe, columns=['column 1'], use_index=True)
        >>>myvega.tabular_data(my_dataframe, columns=['column 1', 'column 2'])

        '''
        
        #Lists
        if isinstance(data, list):
            if append:
                start = self.data[0]['values'][-1]['x'] + 1
                end = len(self.data[0]['values']) + len(data)
            else: 
                start, end = 0, len(data)
                
            default_range = xrange(start, end+1, 1)
            values = [{"x": x, "y": y} for x, y in zip(default_range, data)]
        
        #Dicts   
        if isinstance(data, dict) or isinstance(data, pd.Series):
            values = [{"x": x, "y": y} for x, y in data.iteritems()]
        
        #Dataframes
        if isinstance(data, pd.DataFrame):
            if len(columns) > 1 and use_index: 
                raise ValueError('If using index as x-axis, len(columns)'
                                 'cannot be > 1')
            if use_index or len(columns) == 1: 
                values = [{"x": x[0], "y": x[1][columns[0]]} 
                           for x in data.iterrows()]
            else: 
                values = [{"x": x[1][columns[0]], "y": x[1][columns[1]]} 
                          for x in data.iterrows()]

        if append:
            if not self.data[0]['name']:
                raise ValueError('There is no existing data to append')
            self.data[0]['values'].extend(values)
        else:     
            self.data = []   
            self.data.append({"name": name, "values": values})
        
        if isinstance(values[0]['x'], str) or isinstance(values[0]['y'], str):
            print('Warning: tabular string values require ordinal axes.')
        self.build_vega()                 
                      
class Bar(Vega):
    '''Create a bar chart in Vega grammar'''
    
    def __init__(self):
        '''Build Vega Bar chart with default parameters'''
        super(Bar, self).__init__()
        
        self.scales = [{"name": "x", "type": "ordinal", "range": "width", 
                        "domain": {"data": "table", "field": "data.x"}},
                        {"name": "y", "range": "height", "nice": True, 
                         "domain": {"data": "table", "field": "data.y"}}]
                         
        self.axes = [{"type": "x", "scale": "x"}, {"type": "y", "scale": "y"}]
        
        self.marks = [{"type": "rect", "from": {"data": "table"}, 
                      "properties": {
                                     "enter": {
                                     "x": {"scale": "x", "field": "data.x"},
                                     "width": {"scale": "x", "band": True, 
                                               "offset": -1},
                                     "y": {"scale": "y", "field": "data.y"},
                                     "y2": {"scale": "y", "value": 0}
                                     },      
                                     "update": {"fill": {"value": "#2a3140"}},
                                     "hover": {"fill": {"value": "#a63737"}}
                                     }
                      }]
                      
        self.build_vega()
        
class Area(Bar):
    '''Create an area chart in Vega grammar'''
    
    def __init__(self):
        '''Build Vega Area chart with default parameters'''
        super(Area, self).__init__()
        area_updates = [('remove', 'width', 'marks', 0, 'properties', 'enter'),
                         ('add', 'area', 'marks', 0, 'type'),
                         ('add', 'linear', 'scales', 0, 'type')]
                         
        self.multi_update(area_updates)
        self.build_vega()
        
class Scatter(Bar):
    '''Create a scatter plot in Vega grammar'''
    
    def __init__(self):
        '''Build Vega Scatter chart with default parameters'''
        super(Scatter, self).__init__()
        self.height, self.width = 400, 400
        self.padding = {'top': 40, 'left': 40, 'bottom': 40, 'right': 40}
        scatter_updates = [('remove', 'type', 'scales', 0), 
                           ('add', True, 'scales', 0, 'nice'),
                           ('remove', 'width', 'marks', 0, 'properties', 
                            'enter'),
                           ('remove', 'y2', 'marks', 0, 'properties', 
                            'enter'),
                           ('remove', 'hover', 'marks', 0, 'properties'),
                           ('add', {'value': '#2a3140'}, 'marks', 0, 
                            'properties', 'enter', 'stroke'),
                           ('add', {'value': 0.9}, 'marks', 0, 'properties', 
                            'enter', 'fillOpacity'),
                           ('add', 'symbol', 'marks', 0, 'type')]
                            
        self.multi_update(scatter_updates)
        self.build_vega()
        
class Line(Scatter): 
    '''Create a line plot in Vega grammar'''
    
    def __init__(self):
        '''Build Vega Scatter chart with default parameters'''
        
        pass
        
        #Something still broken- need to do some syntax hunting...
        super(Scatter, self).__init__()
        line_updates = [('remove', 'update', 'marks', 0, 'properties'),
                        ('add', 'line', 'marks', 0, 'type')]

        self.multi_update(line_updates)
        self.build_vega()
        
        
